### PR TITLE
Follow-up to Variant change.

### PR DESCRIPTION
**Pull request by dkorolev**
Follow-up to https://github.com/C5T/Current/pull/605.

### CODE REVIEW

File: TypeSystem/variant.h [hunk 1]
```diff
@@ -245,7 +245,7 @@ struct VariantImpl<NAME, TypeListImpl<TYPES...>> : CurrentVariantImpl<NAME> {
   template <typename... RHS>
   void MoveFrom(VariantImpl<RHS...>&& rhs) {
     if (rhs.object_) {
-      TypeAwareMove mover(rhs.object_, this->object_);
+      TypeAwareMove mover(std::move(rhs.object_), object_);
```
Comment (mzhurovich) [inline TypeSystem/variant.h]: I suppose `std::move` is redundant here.

Comment (dkorolev) [inline TypeSystem/variant.h]: It's not:

```
In file included from test.cc:31:
In file included from ./struct.h:39:
./variant.h:248:21: error: no matching constructor for initialization of
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> > >::TypeAwareMove'
      TypeAwareMove mover(rhs.object_, object_);
                    ^     ~~~~~~~~~~~~~~~~~~~~
./variant.h:84:5: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> >
      >::MoveFrom<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested here
    MoveFrom(std::forward<VariantImpl<RHS...>>(rhs));
    ^
test.cc:815:23: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields> >
      >::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested here
    Variant<Baz, Foo> destination(std::move(source));
                      ^
./variant.h:221:5: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<CurrentSuper>' to
      'std::unique_ptr<CurrentSuper> &&' for 1st argument
    TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
    ^
./variant.h:218:10: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
  struct TypeAwareMove {
         ^
./variant.h:218:10: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
./variant.h:248:21: error: no matching constructor for initialization of
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> >
      >::TypeAwareMove'
      TypeAwareMove mover(rhs.object_, object_);
                    ^     ~~~~~~~~~~~~~~~~~~~~
./variant.h:84:5: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> >
      >::MoveFrom<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested here
    MoveFrom(std::forward<VariantImpl<RHS...>>(rhs));
    ^
test.cc:826:19: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> >
      >::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested here
    destination = std::move(source);
                  ^
./variant.h:221:5: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<CurrentSuper>' to
      'std::unique_ptr<CurrentSuper> &&' for 1st argument
    TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
    ^
./variant.h:218:10: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
  struct TypeAwareMove {
         ^
./variant.h:218:10: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
./variant.h:248:21: error: no matching constructor for initialization of
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >::TypeAwareMove'
      TypeAwareMove mover(rhs.object_, object_);
                    ^     ~~~~~~~~~~~~~~~~~~~~
./variant.h:84:5: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> >
      >::MoveFrom<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested
      here
    MoveFrom(std::forward<VariantImpl<RHS...>>(rhs));
    ^
test.cc:836:19: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> >
      >::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested
      here
    destination = std::move(source);
                  ^
./variant.h:221:5: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<CurrentSuper>' to
      'std::unique_ptr<CurrentSuper> &&' for 1st argument
    TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
    ^
./variant.h:218:10: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
  struct TypeAwareMove {
         ^
./variant.h:218:10: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
./variant.h:248:21: error: no matching constructor for initialization of
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> > >::TypeAwareMove'
      TypeAwareMove mover(rhs.object_, object_);
                    ^     ~~~~~~~~~~~~~~~~~~~~
./variant.h:84:5: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> >
      >::MoveFrom<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested
      here
    MoveFrom(std::forward<VariantImpl<RHS...>>(rhs));
    ^
test.cc:846:32: note: in instantiation of function template specialization
      'current::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> > >,
      current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Foo<current::reflection::DeclareFields>,
      struct_definition_test::CURRENT_STRUCT_IMPL_Baz<current::reflection::DeclareFields> >
      >::VariantImpl<current::reflection::CurrentVariantDefaultName<current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields>
      > >, current::metaprogramming::TypeListImpl<struct_definition_test::CURRENT_STRUCT_IMPL_Bar<current::reflection::DeclareFields> > >' requested
      here
    EXPECT_THROW(destination = std::move(source), IncompatibleVariantTypeException<Bar>);
                               ^
./../3rdparty/gtest/src/../include/gtest/gtest.h:1899:21: note: expanded from macro 'EXPECT_THROW'
  GTEST_TEST_THROW_(statement, expected_exception, GTEST_NONFATAL_FAILURE_)
                    ^
./../3rdparty/gtest/src/../include/gtest/internal/gtest-internal.h:1051:54: note: expanded from macro 'GTEST_TEST_THROW_'
      GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
                                                     ^
./../3rdparty/gtest/src/../include/gtest/internal/gtest-internal.h:1044:44: note: expanded from macro 'GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_'
  if (::testing::internal::AlwaysTrue()) { statement; }
                                           ^
./variant.h:221:5: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<CurrentSuper>' to
      'std::unique_ptr<CurrentSuper> &&' for 1st argument
    TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
    ^
./variant.h:218:10: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
  struct TypeAwareMove {
         ^
./variant.h:218:10: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
4 errors generated.
make: *** [.current/test] Error 1
```

Comment (dkorolev) [inline TypeSystem/variant.h]:
```
g++-5 -std=c++11 -Wall -W -ftemplate-backtrace-limit=0  -ftemplate-depth=10000 -o ".current/test" "test.cc" -pthread -ldl
In file included from struct.h:39:0,
                 from test.cc:31:
variant.h: In instantiation of ‘void current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::MoveFrom(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
variant.h:84:13:   required from ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::VariantImpl(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
test.cc:815:52:   required from here
variant.h:248:47: error: cannot bind ‘std::unique_ptr<current::CurrentSuper>’ lvalue to ‘std::unique_ptr<current::CurrentSuper>&&’
       TypeAwareMove mover(rhs.object_, object_);
                                               ^
variant.h:221:5: note:   initializing argument 1 of ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::TypeAwareMove::TypeAwareMove(std::unique_ptr<current::Cur
     TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
     ^
variant.h: In instantiation of ‘void current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::MoveFrom(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
variant.h:84:13:   required from ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::VariantImpl(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
test.cc:826:17:   required from here
variant.h:248:47: error: cannot bind ‘std::unique_ptr<current::CurrentSuper>’ lvalue to ‘std::unique_ptr<current::CurrentSuper>&&’
       TypeAwareMove mover(rhs.object_, object_);
                                               ^
variant.h:221:5: note:   initializing argument 1 of ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::TypeAwareMove::TypeAwareMove(std::unique_ptr<current::Cur
     TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
     ^
variant.h: In instantiation of ‘void current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::MoveFrom(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
variant.h:84:13:   required from ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::VariantImpl(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
test.cc:836:17:   required from here
variant.h:248:47: error: cannot bind ‘std::unique_ptr<current::CurrentSuper>’ lvalue to ‘std::unique_ptr<current::CurrentSuper>&&’
       TypeAwareMove mover(rhs.object_, object_);
                                               ^
variant.h:221:5: note:   initializing argument 1 of ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::TypeAwareMove::TypeAwareMove(std::unique_ptr<current::Cur
     TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
     ^
variant.h: In instantiation of ‘void current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::MoveFrom(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
variant.h:84:13:   required from ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::VariantImpl(current::VariantImpl<RHS ...>&&) [with RHS = {current::reflectio
test.cc:846:5:   required from here
variant.h:248:47: error: cannot bind ‘std::unique_ptr<current::CurrentSuper>’ lvalue to ‘std::unique_ptr<current::CurrentSuper>&&’
       TypeAwareMove mover(rhs.object_, object_);
                                               ^
variant.h:221:5: note:   initializing argument 1 of ‘current::VariantImpl<NAME, current::metaprogramming::TypeListImpl<REST ...> >::TypeAwareMove::TypeAwareMove(std::unique_ptr<current::Cur
     TypeAwareMove(std::unique_ptr<CurrentSuper>&& from, std::unique_ptr<CurrentSuper>& into)
     ^
make: *** [.current/test] Error 1
```

Comment (mzhurovich) [inline TypeSystem/variant.h]: Hmm. And I actually don't get the idea of using rvalue references here...

**Pull request merged**